src/GraphQL/Server.hs

```diff
@@ -318,44 +325,57 @@ type family RunUnionHandlerType m (a :: Type) :: Type where
 class RunUnion m a where
   runUnion :: RunUnionHandlerType m a -> AST.Selection -> m GValue.Object
 
-instance forall m typeName interfaces fields rest.
+instance forall m object rest.
          ( RunUnion m rest
          , MonadIO m
          , MonadThrow m
-         , RunFields m (RunFieldsType m fields)
-         , KnownSymbol typeName
-         ) => RunUnion m (Object typeName interfaces fields :<|> rest) where
-  runUnion (lh :<|> rh) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection))
-    | typeName == queryTypeName = do
-        result <- buildResolver @m @(Object typeName interfaces fields) lh subSelection
-        -- TODO: See if we can prevent this from happening at compile time.
-        case GValue.toObject result of
-          Nothing -> panic $ "Expected object as result of union query: " <> show result
-          Just object -> pure object
-    | otherwise = runUnion @m @rest rh fragment
-    where typeName = toS (symbolVal (Proxy :: Proxy typeName))
+         , HasGraph m object
+         , HasObjectDefinition object
+         ) => RunUnion m (object :<|> rest) where
+  runUnion (lh :<|> rh) fragment@(AST.SelectionInlineFragment (AST.InlineFragment (AST.NamedType queryTypeName) [] subSelection)) =
+    case getDefinition @object of
```

Comment:
Specifically, as long as this code needs to read out of symbols (either via `getDefinition` or by arbitrarily duplicating some of the logic in `getDefinition`), it will need to handle errors like this. Our options are to panic (which means the compiler does less work for us and we need to be more careful), to mask the error (which we were doing before), or to represent the error explicitly.

By making it explicit, we're giving ourselves to tools to improve it.